File: packages/agent/agent/tools/propose_batch.ts
import { defineTool } from "eve/tools";
import { z } from "zod";

import { categoryInput } from "../lib/schema";

import { setInvoiceFocus } from "../lib/invoice-focus";
import { requireSessionCaller } from "../lib/tenant";
import { writeProposedBatch } from "../lib/write-proposed-batch";

/**
 * Grava um lote proposto como RASCUNHO e roda a conferência — pelo INPUT.
 *
 * Sem gate de aprovação, de propósito: um rascunho não é o razão. Nada
 * auditável foi tocado, e o trigger do banco só protege linhas já confirmadas.
 *
 * É esta separação que faz o cartão de conferência funcionar: "Corrigir" mexe
 * livremente no rascunho, e o gate incide uma vez só, sobre o estado final já
 * revisado — o que torna a decisão binária de aprovar honesta.
 *
 * Para o fluxo normal de upload, prefira `propose_batch_from_extraction`: o
 * extrator já persistiu a extração completa e esta transcrição manual das
 * linhas é exatamente o que a passagem por referência elimina. Esta tool fica
 * para lotes montados na conversa (poucas linhas ditadas pela pessoa) e para
 * repropor com correções que não vieram de extração.
 */
export default defineTool({
  description:
    "Records transactions as a draft batch and checks the sum against the declared total. For a document the extractor just processed, prefer propose_batch_from_extraction (by reference, no retyping). Use this one for batches assembled in conversation. Nothing enters the ledger here — only after the person approves.",
  inputSchema: z.object({
    documentId: z.string().min(1),
    documentKind: z
      .enum(["unknown", "credit_card_invoice", "bank_statement", "invoice_nfe"])
      .optional()
      .describe("Kind of document. Use bank_statement for account statements."),
    issuer: z
      .string()
      .nullable()
      .optional()
      .describe(
        "Who issued the document, as printed on it — 'Nubank', 'Itaú'. Always fill this when the document identifies the issuer: it becomes the title of the verification card.",
      ),
    periodStart: z.string().nullable().optional(),
    periodEnd: z.string().nullable().optional(),
    dueDate: z.string().nullable().optional(),
    declaredTotal: z
      .number()
      .int()
      .nullable()
      .optional()
      .describe("Total declared in the document, IN CENTS. null when the document declares none."),
    declaredSubtotals: z
      .object({
        fees: z.number().int().nullable().optional(),
        purchases: z.number().int().nullable().optional(),
      })
      .nullable()
      .optional()
      .describe(
        'Subtotals from the invoice SUMMARY block, in cents: "IOF de compras internacionais" into `fees`, "Total de compras" into `purchases`. These are what make it possible to say WHERE a discrepancy is.',
      ),
    openingBalance: z.number().int().nullable().optional().describe("Opening bank-statement balance in cents."),
    closingBalance: z.number().int().nullable().optional().describe("Closing bank-statement balance in cents."),
    transactions: z.array(
      z.object({
        date: z.string().regex(/^\d{4}-\d{2}-\d{2}$/),
        originalDescription: z.string().min(1),
        merchant: z.string().nullable().optional(),
        amount: z
          .number()
          .int()
          .describe("IN CENTS, signed: expense > 0, credit (payment/refund) < 0."),
        kind: z
          .enum([
            "purchase",
            "payment",
            "refund",
            "fee",
            "adjustment",
            "income",
            "transfer",
            "card_payment",
            "cash_withdrawal",
          ])
          .optional(),
        installment: z
          .object({ current: z.number().int().positive(), total: z.number().int().positive() })
          .nullable()
          .optional(),
        category: categoryInput(),
        extractionConfidence: z.enum(["alta", "media", "baixa"]),
        page: z.number().int().positive().nullable().optional(),
      }),
    ),
    overwriteEditedDraft: z
      .boolean()
      .optional()
      .describe(
        "Pass true ONLY after the person confirmed discarding an already-corrected draft of this document. Without it, proposing over an edited draft is refused with rascunho_editado.",
      ),
  }),
  async execute(input, ctx) {
    const { tenantId, sessionId } = requireSessionCaller(ctx);

    const result = await writeProposedBatch(tenantId, input);
    if ("error" in result) return result;
    const focused = await setInvoiceFocus(tenantId, sessionId, result.batchId);
    if (!focused) {
      throw new Error("A sessão Eve não estava persistida para guardar o foco da fatura.");
    }

    return {
      batchId: result.batchId,
      status: "proposed" as const,
      documentKind: result.documentKind,
      // O emissor volta no retorno porque é dele que a tela tira o título do
      // cartão. Calculado aqui e não devolvido é o mesmo que não calculado.
      issuer: result.issuer,
      invoiceLabel: result.invoiceLabel,
      periodEnd: result.periodEnd,
      dueDate: result.dueDate,
      ...(input.openingBalance !== undefined ? { openingBalance: input.openingBalance } : {}),
      ...(input.closingBalance !== undefined ? { closingBalance: input.closingBalance } : {}),
      transactionCount: result.transactionCount,
      checksum: result.checksum,
      ...(result.statementBalance === undefined
        ? {}
        : { statementBalance: result.statementBalance }),
      ...(result.duplicateSuspects !== undefined
        ? {
            duplicateSuspects: result.duplicateSuspects,
            duplicateNote:
              "Linhas com mesma data, valor e comerciante já CONFIRMADAS vindas de outro documento — possível dupla contagem (fatura parcial + fechada). AVISE a pessoa antes de abrir o commit; ela decide com os ids em mãos.",
          }
        : {}),
      next: "If the verification looks good, call `commit_batch` to open the decision. Do not ask in prose whether you may record it.",
    };
  },
});

File: services/agent/clara/tools/propose_batch.py
"""`propose_batch` e `propose_batch_from_extraction` — `@tool` de fronteira
sobre `write_proposed_batch`. Porta de `agent/tools/propose_batch.ts` e
`propose_batch_from_extraction.ts`.
"""

from __future__ import annotations

from typing import Any

from agno.run.base import RunContext
from agno.tools import tool
from pydantic import BaseModel
from sqlalchemy import select

from clara.db.models import ExtractionStaging
from clara.db.tenant_scope import for_tenant
from clara.ledger.types import Confidence, EntryKind
from clara.tools.errors import not_found
from clara.tools.serialize import to_tool_dict
from clara.tools.tenant import require_tenant_caller
from clara.tools.write_proposed_batch import (
    ProposedBatchWriteInput,
    ProposedTransactionInput,
)
from clara.tools.write_proposed_batch import write_proposed_batch as _write


class InstallmentInput(BaseModel):
    current: int
    total: int


class TransactionInput(BaseModel):
    date: str
    original_description: str
    amount: int
    extraction_confidence: Confidence
    merchant: str | None = None
    kind: EntryKind = "purchase"
    installment: InstallmentInput | None = None
    category: str | None = None
    page: int | None = None


def _to_domain(t: TransactionInput) -> ProposedTransactionInput:
    return ProposedTransactionInput(
        date=t.date, original_description=t.original_description, amount=t.amount,
        extraction_confidence=t.extraction_confidence, merchant=t.merchant, kind=t.kind,
        installment_current=t.installment.current if t.installment else None,
        installment_total=t.installment.total if t.installment else None,
        category=t.category, page=t.page,
    )


@tool(
    name="propose_batch",
    description=(
        "Proposes a draft batch assembled in conversation — a few lines dictated by the "
        "person. For a document the extractor read, use propose_batch_from_extraction "
        "instead; never retype the extractor's transactions here."
    ),
)
def propose_batch(
    run_context: RunContext,
    document_id: str,
    transactions: list[TransactionInput],
    document_kind: str | None = None,
    issuer: str | None = None,
    period_start: str | None = None,
    period_end: str | None = None,
    due_date: str | None = None,
    declared_total: int | None = None,
    opening_balance: int | None = None,
    closing_balance: int | None = None,
    overwrite_edited_draft: bool = False,
) -> dict[str, Any]:
    caller = require_tenant_caller(run_context)
    write_input = ProposedBatchWriteInput(
        document_id=document_id,
        transactions=[_to_domain(t) for t in transactions],
        document_kind=document_kind,  # type: ignore[arg-type]
        issuer=issuer, period_start=period_start, period_end=period_end, due_date=due_date,
        declared_total=declared_total, opening_balance=opening_balance,
        closing_balance=closing_balance, overwrite_edited_draft=overwrite_edited_draft,
    )
    with for_tenant(caller.tenant_id) as session:
        result = _write(session, caller.tenant_id, write_input)
    return to_tool_dict(result)


@tool(
    name="propose_batch_from_extraction",
    description=(
        "Proposes a draft batch from a persisted extraction (extraction_id from "
        "save_extraction's receipt). This is the reference path — the coordinator never "
        "retypes the transactions the extractor read."
    ),
)
def propose_batch_from_extraction(
    run_context: RunContext,
    extraction_id: str,
    overwrite_edited_draft: bool = False,
) -> dict[str, Any]:
    caller = require_tenant_caller(run_context)

    with for_tenant(caller.tenant_id) as session:
        staging = session.execute(
            select(ExtractionStaging).where(
                ExtractionStaging.id == extraction_id,
                ExtractionStaging.tenant_id == caller.tenant_id,
            )
        ).scalar_one_or_none()

        if staging is None:
            return to_tool_dict(
                not_found(
                    "extracao_nao_encontrada", f"Nenhuma extração com o id {extraction_id}.",
                    hint="Chame save_extraction de novo, ou confira o extraction_id do recibo.",
                )
            )

        payload = staging.payload
        write_input = ProposedBatchWriteInput(
            document_id=payload["document_id"],
            document_kind=payload.get("document_kind"),
            issuer=payload.get("issuer"),
            period_start=payload.get("period_start"),
            period_end=payload.get("period_end"),
            due_date=payload.get("due_date"),
            declared_total=payload.get("declared_total"),
            declared_subtotals=payload.get("declared_subtotals"),
            opening_balance=payload.get("opening_balance"),
            closing_balance=payload.get("closing_balance"),
            overwrite_edited_draft=overwrite_edited_draft,
            transactions=[
                ProposedTransactionInput(
                    date=t["date"], original_description=t["original_description"],
                    amount=t["amount"], extraction_confidence=t["extraction_confidence"],
                    merchant=t.get("merchant"), kind=t.get("kind", "purchase"),
                    installment_current=(t.get("installment") or {}).get("current"),
                    installment_total=(t.get("installment") or {}).get("total"),
                    category=t.get("category"), page=t.get("page"),
                )
                for t in payload["transactions"]
            ],
        )
        result = _write(session, caller.tenant_id, write_input)
    return to_tool_dict(result)
